### PR TITLE
feat(Core/QuestTracker): implement quest tracker manager

### PR DESCRIPTION
**Pull request by Winfidonarleyan**
## CHANGES PROPOSED:
- Implement queue 
- Add separate class
- Rework configuration for system
```ini
#
#    QuestTracker.Enable
#        Description: Store data in the database about quest completion and abandonment to help finding bugged quests.
#        Default:     0  - (Disabled)
#                     1  - (Enabled)
#
#    QuestTracker.Queue.Enable
#        Description: Enable queue
#        Default:     0  - (Disabled)
#                     1  - (Enabled)
#
#    QuestTracker.Queue.Delay
#        Description: Delay for update queue in seconds. 1 seconds minimun
#        Default:     10
#

QuestTracker.Enable = 0
QuestTracker.Queue.Enable = 0
QuestTracker.Queue.Delay = 10
```

## ISSUES ADDRESSED:
- #4005

## Tests Performed:
- https://gist.github.com/Winfidonarleyan/e952b58c701f272ac62faa303253b9bd

## How to Test the Changes:
- Download this module https://github.com/WarheadCore/mod-quest-tracker-test-ac
- Compile core with this PR
- Add macro in-game
```
.questt 500 0
.questt 500 1
.questt 500 2
.questt 500 3
```
- Check worldserver console
- Check DB table `quest_tracker`

## Known Issues and TODO List:
- [ ] 

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Winfidonarleyan** [comment]
@mpfans try this

File: src/common/Utilities/TaskScheduler.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/common/Utilities/TaskScheduler.cpp:2]: please adapt this to AzerothCore, it's ok to mention WarheadCore as well but use the existing format

File: src/common/Utilities/TaskScheduler.h [hunk 2]
```diff
@@ -0,0 +1,647 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/common/Utilities/TaskScheduler.h:2]: same

File: src/server/game/Quests/QuestTracker.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/server/game/Quests/QuestTracker.cpp:2]: same

File: src/server/game/Quests/QuestTracker.cpp [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+{
+    static QuestTracker instance;
+    return &instance;
+}
+
+void QuestTracker::InitSystem()
+{
+    if (!sWorld->getBoolConfig(CONFIG_QUEST_ENABLE_QUEST_TRACKER))
+    {
+        sLog->outString(">> System disabled");
```
Comment (FrancescoBorzi) [inline src/server/game/Quests/QuestTracker.cpp:42]: let's use a more meaningful message here, so we know where it comes from. Example:

```
">> The QuestTracker System is disabled"
```

File: src/server/game/Quests/QuestTracker.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
+/*
+ * This file is part of the WarheadCore Project. See AUTHORS file for Copyright information
```
Comment (FrancescoBorzi) [inline src/server/game/Quests/QuestTracker.h:2]: same as above

File: src/server/game/World/World.cpp [hunk 6]
```diff
@@ -2028,6 +2029,9 @@ void World::SetInitialWorldSettings()
     mgr = ChannelMgr::forTeam(TEAM_HORDE);
     mgr->LoadChannels();
 
+    sLog->outString("Start QuestTracker system...");
```
Comment (FrancescoBorzi) [inline src/server/game/World/World.cpp:2032]: better `Initialising QuestTracker system...`

**Review by FrancescoBorzi — CHANGES_REQUESTED** [review]
@Winfidonarleyan good job, I left some remarks

**Comment by mpfans** [comment]
good job! not delay...but there are some error messages

**Comment by mpfans** [comment]
1.Completion of quest prior to acceptance of commitments is not documented.
2.New quest are recorded

3.abandon quest are recorded

4.abandon  quest and get the record again

5.quest complete BY GM are recorded


6.quest 10113  no record,There is no record of get and complete
quest 10113  is before get the request 9857

After testing, when receiving a large number of quests at the same time, some quests will be missed

**Comment by Winfidonarleyan** [comment]
Hmm, okay, i try fix

**Comment by Winfidonarleyan** [comment]
@mpfans Try retest with https://github.com/azerothcore/azerothcore-wotlk/pull/4112/commits/f27b448c157b4378d4eea62dfcba3ef110bb4342

**Comment by mpfans** [comment]
> @mpfans Try retest with [f27b448](https://github.com/azerothcore/azerothcore-wotlk/commit/f27b448c157b4378d4eea62dfcba3ef110bb4342)

How can I test it?

**Comment by Winfidonarleyan** [comment]
Retest this PR. Do it exactly the same way

**Comment by mpfans** [comment]
Normal work, about an hour after running, the server will crash.Can you tell me how to get the crash log?

**Comment by Winfidonarleyan** [comment]
https://trinitycore.atlassian.net/wiki/spaces/tc/pages/2130173/GDB

**Comment by mpfans** [comment]
i use 'git reset --hard f27b448' and then recompile

**Comment by mpfans** [comment]
1.on master branch 
2.git reset --hard 039e143
3.git checkout pr-4112
4.git pull origin pull/4112/head
5.cd build
6.cmake ........
7.make -j32
8.
image

**Comment by mpfans** [comment]
> https://trinitycore.atlassian.net/wiki/spaces/tc/pages/2130173/GDB



> Normal work, about an hour after running, the server will crash.Can you tell me how to get the crash log?

https://github.com/azerothcore/azerothcore-wotlk/issues/4225

**Comment by Winfidonarleyan** [comment]
Revert this commit and try retest

**Comment by mpfans** [comment]
> Revert [b276162](https://github.com/azerothcore/azerothcore-wotlk/commit/b2761626fec1d454acb58f08c0e0e3d8c9b83e47) and try retest

ok.Please wait my feedback

**Comment by mpfans** [comment]
image
still compile error

**Comment by Winfidonarleyan** [comment]
Ah, my bad, sorry, I fix it today

**Comment by Winfidonarleyan** [comment]
@mpfans try

**Comment by mpfans** [comment]
compile  ok

not work... game delay....bbout 500 players started to delay

File: src/server/game/Quests/QuestTracker.cpp [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        return;
+    }
+
+    scheduler.Schedule(10s, [this](TaskContext context)
```
Comment (Winfidonarleyan) [inline src/server/game/Quests/QuestTracker.cpp:48]: @mpfans Try to change to `20s` or `1min`
```cpp
scheduler.Schedule(1min, [this](TaskContext context)
```

**Comment by Winfidonarleyan** [comment]
@mpfans If this system disable core normal work? no delay?

**Comment by mpfans** [comment]
> compile ok
> 
> not work... game delay....bbout 500 players started to delay

Quests.EnableQuestTracker = 1 delay
Quests.EnableQuestTracker = 0 not delay

**Comment by mpfans** [comment]
> @mpfans If this system disable core normal work? no delay?
> Does the delay appear if the system is enable?

i recompile for 'scheduler.Schedule(1min, [this](TaskContext context)' , please wait feedback

**Comment by Winfidonarleyan** [comment]
Okay, wait

**Comment by mpfans** [comment]
> Okay, wait
still delayed

**Comment by Winfidonarleyan** [comment]
No need to rush, the main thing is the result

**Comment by mpfans** [comment]
https://github.com/azerothcore/azerothcore-wotlk/pull/4112/commits/0e89e09de224b85d91a036dd12470188594171b0   no delay,but same quest not record

https://github.com/azerothcore/azerothcore-wotlk/pull/4112/commits/f27b448c157b4378d4eea62dfcba3ef110bb4342 have delay

**Comment by Winfidonarleyan** [comment]
Hmmm

**Comment by Winfidonarleyan** [comment]
Try now, i change to 1 trans for updates

**Comment by mpfans** [comment]
image
compile error

**Comment by Winfidonarleyan** [comment]
fix

**Comment by mpfans** [comment]
image
compile error

**Comment by Winfidonarleyan** [comment]
again

**Comment by mpfans** [comment]
> again

compile ok... please wait for another test

**Comment by mpfans** [comment]
image

File: src/server/game/Quests/QuestTracker.cpp [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+void QuestTracker::InitSystem()
+{
+    if (!sWorld->getBoolConfig(CONFIG_QUEST_ENABLE_QUEST_TRACKER))
+    {
+        sLog->outString(">> System disabled");
+        return;
+    }
+
+    scheduler.Schedule(30s, [this](TaskContext context)
```
Comment (FrancescoBorzi) [inline src/server/game/Quests/QuestTracker.cpp:48]: maybe this parameter can be configurable?

**Comment by Winfidonarleyan** [comment]
@mpfans Try to increase 
```ini
CharacterDatabase.WorkerThreads
```

**Comment by mpfans** [comment]
> @mpfans Try to increase
> 
> ```ini
> CharacterDatabase.WorkerThreads
> ```

image
From the use of AC, has been set to 10

**Comment by Viste** [comment]
> image
> 
> There are still some quests not recorded

can you show full DB log, especially before this lock seems somthing lock table/DB need to understand what.
And it's most likely not the quest tracker itself, but something else. It just exposes the problem. And the delay form original patch can appear because of this lock

**Comment by mpfans** [comment]
> > image
> > There are still some quests not recorded
> 
> can you show full DB log, especially before this lock seems somthing lock table/DB need to understand what.
> And it's most likely not the quest tracker itself, but something else. It just exposes the problem. And the delay form original patch can appear because of this lock

[DBError.log](https://github.com/azerothcore/azerothcore-wotlk/files/5786958/DBError.log)

**Comment by Viste** [comment]
Okay. One more thing I forgot to ask.
Are there any queries to the characters database on the side of your web page?

**Comment by mpfans** [comment]
> Okay. One more thing I forgot to ask.
> Are there any queries to the characters database on the side of your web page?

yes , we have

**Comment by FrancescoBorzi** [comment]
> > Okay. One more thing I forgot to ask.
> > Are there any queries to the characters database on the side of your web page?
> 
> yes , we have

could you please test with external DB queries disabled? so we can make sure they are not the cause of the issue

**Comment by mpfans** [comment]
All web queries are closed

and then
image

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan any plans to finish this before it gets more and more conflicts? :)

**Comment by Winfidonarleyan** [comment]
I don't know if this pr.

**Comment by Winfidonarleyan** [comment]
Hmmmmmmmm, I have no problem with transactions

**Comment by Winfidonarleyan** [comment]
Okay, testers, need help

**Comment by Winfidonarleyan** [comment]
Implement module for test this PR and add info `How to Test the Changes`

**Comment by Winfidonarleyan** [comment]
Example DB log https://gist.github.com/Winfidonarleyan/e952b58c701f272ac62faa303253b9bd

**Comment by Winfidonarleyan** [comment]
@mpfans Try check now

**Comment by Winfidonarleyan** [comment]
Added info. Example
```
21:05:20 > QuestTracker: Start Execute...
21:05:20 > QuestTracker: Execute 'CHAR_INS_QUEST_TRACK' (7500)
21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_COMPLETE_TIME' (7500)
21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_ABANDON_TIME' (7500)
21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_GM_COMPLETE' (7500)
21:05:20 > QuestTracker: Execute end in 78 ms
```

**Comment by FrancescoBorzi** [comment]
> Added info. Example
> 
> ```
> 21:05:20 > QuestTracker: Start Execute...
> 21:05:20 > QuestTracker: Execute 'CHAR_INS_QUEST_TRACK' (7500)
> 21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_COMPLETE_TIME' (7500)
> 21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_ABANDON_TIME' (7500)
> 21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_GM_COMPLETE' (7500)
> 21:05:20 > QuestTracker: Execute end in 78 ms
> ```

shouldn't them be only at debug level? or they are just temporary?

**Comment by Winfidonarleyan** [comment]
> > Added info. Example
> > ```
> > 21:05:20 > QuestTracker: Start Execute...
> > 21:05:20 > QuestTracker: Execute 'CHAR_INS_QUEST_TRACK' (7500)
> > 21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_COMPLETE_TIME' (7500)
> > 21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_ABANDON_TIME' (7500)
> > 21:05:20 > QuestTracker: Execute 'CHAR_UPD_QUEST_TRACK_GM_COMPLETE' (7500)
> > 21:05:20 > QuestTracker: Execute end in 78 ms
> > ```
> 
> shouldn't them be only at debug level? or they are just temporary?

For test PR only, After testing, I will rework pr and remove this

**Comment by mpfans** [comment]
srrrrrry...
I update to c9b46beae276d8990cea7935609d4a8461d40bf3

image
Can you update the latest version?

**Comment by Winfidonarleyan** [comment]
@mpfans try

**Comment by mpfans** [comment]
image

**Comment by Winfidonarleyan** [comment]
@Winfidonarleyan in ac discord

**Comment by Winfidonarleyan** [comment]
And try hard reset to 3cd88101243e7bf545139d21ff014bf0ef4c7000

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan merge conflicts

**Comment by Winfidonarleyan** [comment]
So, after my test, this PR `DOESN'T` solve the problem @mpfans  (#4005). To `SOLVE` the problem need `MORE MORE MORE` rework DB, VERY MORE.

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan ok, marking this as WIP. Please, let us know when it's ready to be reviewed/tested.

**Comment by Winfidonarleyan** [comment]
? This PR end

**Comment by FrancescoBorzi** [comment]
> ? This PR end

> So, after my test, this PR DOESN'T solve the problem @mpfans (#4005). To SOLVE the problem need MORE MORE MORE rework DB, VERY MORE.

**Comment by Viste** [comment]
билд не падай вставай, не болей. бля ну падает и падает пиздец

**Comment by FrancescoBorzi** [comment]
@Winfidonarleyan so, if I understood correctly, this PR does not solve mpfans' issue, but it's still something we want to merge?

File: src/server/worldserver/worldserver.conf.dist [hunk 9]
```diff
@@ -1303,12 +1303,24 @@ Instance.ResetTimeHour = 4
...
 
 #
-#    Quests.EnableQuestTracker
+#    QuestTracker.Enable
 #        Description: Store data in the database about quest completion and abandonment to help finding bugged quests.
 #        Default:     0  - (Disabled)
 #                     1  - (Enabled)
+#
+#    QuestTracker.Queue.Enable
+#        Description: Enable queue
```
Comment (FrancescoBorzi) [inline src/server/worldserver/worldserver.conf.dist:1312]: what does `Enable queue` mean? @Winfidonarleyan can you please add a more detailed description?

Comment (Winfidonarleyan) [reply src/server/worldserver/worldserver.conf.dist:1312]: Need help for this desc :D

**Comment by Winfidonarleyan** [comment]
Close this pr, after #5611 i remake this

**Pull request closed without merging**